app/scripts/migrations/137.ts

```diff
@@ -0,0 +1,318 @@
+      new Error(
+        `Migration ${version}: typeof state.NetworkController.selectedNetworkClientId is ${typeof selectedNetworkClientId}`,
+      ),
+    );
+    return state;
+  }
+  if (!isObject(networkConfigurationsByChainId)) {
+    global.sentry?.captureException(
+      new Error(
+        `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId is ${typeof networkConfigurationsByChainId}`,
+      ),
+    );
+    return state;
+  }
+  if (!isObject(domains)) {
+    global.sentry?.captureException(
+      new Error(
+        `Migration ${version}: typeof state.SelectedNetworkController.domains is ${typeof domains}`,
+      ),
+    );
+    return state;
+  }
+
+  const getChainIdForNetworkClientId = (
+    networkClientId: string,
+    propertyName: string,
+  ): string | undefined => {
+    for (const [chainId, networkConfiguration] of Object.entries(
+      networkConfigurationsByChainId,
+    )) {
+      if (!isObject(networkConfiguration)) {
+        global.sentry?.captureException(
+          new Error(
+            `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"] is ${typeof networkConfiguration}`,
+          ),
+        );
+        return undefined;
+      }
+      if (!Array.isArray(networkConfiguration.rpcEndpoints)) {
+        global.sentry?.captureException(
+          new Error(
+            `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"].rpcEndpoints is ${typeof networkConfiguration.rpcEndpoints}`,
+          ),
+        );
+        return undefined;
+      }
+      for (const rpcEndpoint of networkConfiguration.rpcEndpoints) {
+        if (!isObject(rpcEndpoint)) {
+          global.sentry?.captureException(
+            new Error(
+              `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"].rpcEndpoints[] is ${typeof rpcEndpoint}`,
+            ),
+          );
+          return undefined;
+        }
+        if (rpcEndpoint.networkClientId === networkClientId) {
+          return chainId;
+        }
+      }
+    }
+
+    const builtInChainId = BUILT_IN_NETWORKS.get(networkClientId);
+    if (!builtInChainId) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: No chainId found for ${propertyName} "${networkClientId}"`,
+        ),
+      );
+    }
+    return builtInChainId;
+  };
+
+  const currentChainId = getChainIdForNetworkClientId(
+    selectedNetworkClientId,
+    'selectedNetworkClientId',
+  );
+  if (!currentChainId) {
+    return state;
+  }
+
+  for (const [origin, subject] of Object.entries(subjects)) {
+    if (!isObject(subject)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid subject for origin "${origin}" of type ${typeof subject}`,
+        ),
+      );
+      continue;
+    }
+
+    const { permissions } = subject as {
```

Comment:
The type cast here represents an unverified assumption about the state shape. We can replace this with type guards to verify the state shape, avoiding the need for casts. You've already taken this approach in most of the migration, was there something in the way of validating this particular piece?

Reply:
I briefly tried this locally, and TypeScript was being a bit stubborn about the type inference until I pulled out `permissions[PermissionNames.eth_accounts]` into its own local variable. It doesn't apply type inference to properties referenced by name for some reason.

`basePermission` gave me a bit of trouble still, but I'm not totally sure what that was intended to do. It looks like maybe it was an attempt at preserving the permission ID and/or date? Perhaps that would be simpler by capturing those independently, easier to type that way.